### DISCUSSION TITLE
Does pyright support nested protocol?

### DISCUSSION THREAD

**Discussion by raceychan** [microsoft/pyright#6580] [Q&A]
pylance would complain if i use nested protocl(a protocol class as an attribute of another protocol class)
![image](https://github.com/microsoft/pyright/assets/75417963/bd8f74ed-0079-4574-b8ad-0bacf5c48c41)

my use case is something like:
```python
import typing as ty


class Abstract(ty.Protocol):
    class IInner(ty.Protocol):
        something: str

    inner: IInner


class Concrete:
    class Inner:
        somethings: str

    inner: Inner


concrete: Abstract = Concrete()
```
I tried refactor it in this way, but it does not work still.

```python
import typing as ty


class IInner(ty.Protocol):
    something: str


class Abstract(ty.Protocol):
    inner: IInner


class Inner:
    somethings: str


class Concrete:
    inner: Inner


concrete: Abstract = Concrete()

```

I searched for similar issues/questions, and the closest issue i found is 
https://github.com/microsoft/pyright/issues/3207

but it looks a whole lot more complicated than my use case

**Comment by erictraut** [comment]
No, this isn't how protocols work in the Python type system. Supporting this would require a modification to the typing spec. Mypy issues the same error in this case.

**Reply by raceychan** [reply]
so protocol would not apply to cases where we have composed classs where a class use other non-bulitin classes as its members?

**Reply by erictraut** [reply]
You could do this:

```python
from typing import Protocol

class IInner(Protocol):
    something: str

class Abstract(Protocol):
    inner: IInner

class Concrete:
    class Inner:
        something: str

    inner: IInner

concrete: Abstract = Concrete()
```

You cannot include `IInner` within the protocol definition because that indicates that any type compatible with the protocol must have a class-level symbol named `IInner`, which isn't your intention. So you'd need to pull the `IInner` protocol definition outside of the `Abstract` definition.

The `inner` variable is mutable, so its type is considered invariant. That means `Concrete.inner` cannot have the type `Inner` if you want it to match the protocol. It must be defined as `IInner`.

----

If you have questions specifically about pyright, I'm happy to answer them here in this discussion forum, but if you have general questions about the Python type system, I recommend posting to https://discuss.python.org/c/typing/32. That forum is frequented by many people who can answer general typing questions, and the answers will also benefit a wider audience.
